Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,280 @@
+                          (r+1, c-1),(r+1, c),(r, c+1),(r+1, c+1)]
+
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result if result else None
+
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        word = word.lower()
+        return self.findString(word) is not None
+
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        but can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: list(Point) or None
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints:
+            # we are here only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        if not path:
+            return False
+        return [p.getValue() for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: list(Point) or None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if not possible_ways:
+            # we at the same time pop the last element from path and append it to list visited
+            visited.append(path.pop(-1))
+            if path:
+                self._findPathRecursively(path[-1], word, path, visited)
+
+        for neighbor in possible_ways:
+            path.append(neighbor)
+            self._findPathRecursively(neighbor, word, path, visited)
+
+        return path
+
+    def findPossibleWays(self, path, point, word, visited):
+        """
+        Return the list of Points that can be the next element in the word.
+        :return: list(Point)
+        """
+        neighbors = self.findNeighborsOfPoint(point.getCoordinates())
+        ways = []
+        for n in neighbors:
+            # to not to use one Point twice -->
+            if self._pathIsPartOfWord(path + [n], word) and n not in path:
+                # it is of no use to go in the Point where we had been already -->
+                if n not in visited:

```

Comment:
It looks like it is generally not required to pass along the `visited` list because the `path` also contains all the visited nodes.